Add @user-groups support

Aerex opened the pull request:
Add @user-groups with tab-completion. @user-groups will be unfurl into format `<!subteam^{ID}|handle> message` before sending message via `linktext` method

* Implementation includes unit test

torhve commented:
I like this, but shouldn't this PR also add support for notification to usergroups that you are in?

Aerex commented:
@torhve That is a good point. I will start add that into the PR.

trygveaa commented on wee_slack.py:
```diff
@@ -917,6 +916,23 @@ def emoji_completion_cb(data, completion_item, current_buffer, completion):
+@utf8_decode
+def usergroups_completion_cb(data, completion_item, current_buffer, completion):
+    """
+    Adds all :-prefixed usergroups to completion list
```
`:-prefixed` isn't right for usergroups.

Aerex replied:
typo error I will fixed that.

trygveaa commented on wee_slack.py:
```diff
@@ -917,6 +916,23 @@ def emoji_completion_cb(data, completion_item, current_buffer, completion):
+        return w.WEECHAT_RC_OK
+    subteam_ids = current_channel.team.subteams.viewkeys()
+    for subteam_id in subteam_ids:
+        subteam = current_channel.team.subteams.get(subteam_id, None)
```
To loop through a dict getting the value you can use:
```
for subteam in current_channel.team.subteams.values():
```

trygveaa commented on wee_slack.py:
```diff
@@ -1021,14 +1037,34 @@ def should_try(self):
+       if compare_str == self.subteam_id :
+           return True
+       else:
+           return False
```
Just do:
```
return compare_str == self.subteam_id
```

Aerex replied:
Yeah not sure why I did that. I will fix that

trygveaa replied:
You mean remove if/else from that too? Sure, go ahead.

Though, when I think about it, these methods don't really make sense. `__eq__` should check if it equals another instance of the same class, not compare with a string. I wonder if they are used anywhere.

trygveaa commented on wee_slack.py:
```diff
@@ -1062,6 +1099,14 @@ def __init__(self, eventrouter, token, websocket_url, team_info, nick, myidentif
+        for s in self.subteams.keys():
+            subteam = self.subteams.get(s, None)
+            if subteam:
+                self.subteam_handles_to_id[subteam.handle] = subteam.identifier
```
This should be a method, like `get_username_map`, in case `self.subteams` changes.

trygveaa commented on wee_slack.py:
```diff
@@ -917,6 +916,23 @@ def emoji_completion_cb(data, completion_item, current_buffer, completion):
+    for subteam_id in subteam_ids:
+        subteam = current_channel.team.subteams.get(subteam_id, None)
+        if subteam:
+            w.hook_completion_list_add(completion, "@" + subteam.handle, 0, w.WEECHAT_LIST_POS_SORT)
```
Even though user groups are not nicks, I think we should set `nick_completion` to 1, because they are used like nicks. This will cause a colon to be inserted after the user group if it is completed on the start of the line.

Aerex commented:
@trygveaa I have added [subteam_created](https://api.slack.com/events/subteam_created), [subteam_updated](https://api.slack.com/events/subteam_updated), [subteam_self_added](https://api.slack.com/events/subteam_self_added), and [subteam_self_removed](https://api.slack.com/events/subteam_self_removed) event as process handlers. The [subteam_members_changed](https://api.slack.com/events/subteam_members_changed) event didn't seem useful to me compared to the others but if you feel like we should add let me know. 

Also, I have included a usergroups command that will print out the users for a given usergroup given the handler as an argument. Usage is described in the documentation.  

I was going to add some test for handling the `usersgroups.users.list` but it would be more involved that I thought it would be worth. You don't seem to have any tests for the other request handlers. 

Let know if you have any comments, questions, or concerns. Thanks

trygveaa commented on wee_slack.py:
```diff
@@ -2957,6 +3100,8 @@ def linkify_text(message, team, channel):
             if named[1] in ["group", "channel", "here"]:
                 message[item[0]] = "<!{}>".format(named[1])
+            elif named[1] in list(channel.team.subteam_handles_to_id.viewkeys()):
+                message[item[0]] = "<!subteam^{}|{}>".format(channel.team.subteam_handles_to_id[named[1]], named[1])
```
If you write e.g. colon after the group, it will be lost because `named[2]` is not included. I see this was an issue with group/channel/here too, I've fixed that in master now.

trygveaa commented on wee_slack.py:
```diff
@@ -3488,6 +3635,34 @@ def command_users(data, current_buffer, args):
+            team.buffer_prnt("    {:<25}(@{})".format(subteam.name, subteam.handle))
+    else:
+        w.prnt('', 'ERROR: Unknown usergroup handler: {}'.format(args))
+        w.prnt('', 'Usage: /slack usergroups <handler>')
```
Don't print usage info when the command syntax was correct. The first message is sufficient.

trygveaa commented on wee_slack.py:
```diff
@@ -3241,6 +3386,8 @@ def tag(tagset, user=None, thread=False, muted=False):
         "team_info": {"no_highlight", "log3"},
         "team_message": {"irc_privmsg", "notify_message", "log1"},
+        # messages from a subgroup/usergroup 
+        "usergroup":{"irc_privmsg", "notify_message", "log1"},
```
I can't see that you use this tag anywhere?

Aerex commented on docs/Commands.md:
```diff
@@ -173,3 +173,19 @@ Uploads a file to the current buffer.
+List the users in a usergroup using handler (omit @)
+
+
+
```
I did run the script however I did not notice that it was added when I staged my files.

Aerex commented on wee_slack.py:
```diff
@@ -2639,6 +2725,63 @@ def process_message(message_json, eventrouter, store=True, download=True, **kwar
+    team.subteams[new_subteam_info.get('id')] = SlackSubteam(team.identifier, **new_subteam_info)
+
+    for message in messages:
+        team.buffer_prnt(message, message=True)
```
Fair enough. I wasn't certain myself so I added as much as I could just in case. I remove of the usergroups notification
events except when the usergroup handle changes. A user can enable or disable that option in the configuration. I think
it useful if the handle changes so you can keep using without interruptions.

Aerex commented:
@trygveaa Thanks for the review. I added your suggestions to my pull request. I added some two comments from your
responses. I will reiterate one comment here. In regards to the notification, I removed all of the notification
usergroup except for when the handle has changed. You can enable or disable the notification in the new setting 
**notify_usergroup_updated**. Let me know if you have any questions or concerns. Thanks

trygveaa commented on wee_slack.py:
```diff
@@ -2973,6 +3046,8 @@ def linkify_text(message, team):
             if named[1] in ["group", "channel", "here"]:
                 message[item[0]] = "<!{}>{}".format(named[1], named[2])
+            elif named[1] in list(usergroups.viewkeys()):
+                message[item[0]] = "<!subteam^{}|{}>".format(usergroups[named[1]], named[1])
```
Did you see my comment about including `named[2]` here?

> If you write e.g. colon after the group, it will be lost because named[2] is not included. I see this was an issue with group/channel/here too, I've fixed that in master now.

Aerex replied:
I did but I must overlooked something when I was testing.

trygveaa commented on wee_slack.py:
```diff
@@ -3026,6 +3102,12 @@ def unfurl_ref(ref, ignore_alt_text, auto_link_display):
+                if ref.split('|')[1].startswith('@'):
+                    handle = ref.split('|')[1][1:]
+                else:
+                    handle = ref.split('|')[1]
```
Hm, so the handle in `<!subteam^U2147483697|@group>` doesn't include `@` for you?

Aerex replied:
In a channel I found some of the usergroups printing the `@` symbol twice like `@@dev`. I have only seen this in old
messages. These two lines account for a possible extra `@` sign and removes them. Maybe it just me.

trygveaa replied:
Ah, the web-client includes @ in the alt-text when creating a message, while wee-slack does not. I changed linkify_text to include @ in wee-slack as well.

trygveaa commented on wee_slack.py:
```diff
@@ -927,6 +930,21 @@ def emoji_completion_cb(data, completion_item, current_buffer, completion):
+    if current_channel is None:
+        return w.WEECHAT_RC_OK
+    for subteam in current_channel.team.subteams.values():
+        if subteam:
```
What is the reason for this if?

trygveaa commented on wee_slack.py:
```diff
@@ -2649,6 +2699,28 @@ def process_message(message_json, eventrouter, store=True, download=True, **kwar
+    team = kwargs['team']
+    usergroups = team.generate_usergroup_map()
+    new_subteam_info = subteam_json['subteam']
+    messages = []
```
This isn't used anymore.

trygveaa commented on wee_slack.py:
```diff
@@ -2649,6 +2699,28 @@ def process_message(message_json, eventrouter, store=True, download=True, **kwar
+
+    if config.notify_usergroup_handle_updated and current_subteam_info.handle != new_subteam_info['handle']:
+        usergroups[new_subteam_info['handle']] = new_subteam_info.get('id')
+        template = '{name} has updated its handle to @{handle} in team {team}'
```
Since we don't use the user group name in wee-slack anywhere, I think it would be more convenient to write `User group @{old_handle} has ...`.

trygveaa commented on _pytest/test_processsubteamupdated.py:
```diff
@@ -0,0 +1,30 @@
+    eventrouter.teams[t].ws = socket
+    datafiles = glob.glob("_pytest/data/websocket/1483975206.59-subteam_updated.json")
+
+    with patch('wee_slack.SlackTeam.buffer_prnt') as fake_buffer_prnt:
```
Using patch here isn't necessary anymore.

Aerex commented:
@trygveaa I added a comment in regards to this comment 
> Hm, so the handle in <!subteam^U2147483697|@group> doesn't include @ for you?

trygveaa commented:
Thanks!
